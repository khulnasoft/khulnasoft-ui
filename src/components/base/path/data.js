export const mockPathItems = [
  {
    title: 'First',
    metric: '1d',
  },
  {
    title: 'Second',
    metric: '2d',
  },
  {
    title: 'Third',
    metric: '3d',
  },
  {
    title: 'Fourth',
    metric: '4d',
  },
  {
    title: 'Fifth',
    metric: '5d',
  },
  {
    title: 'Sixth',
    metric: '6d',
  },
  {
    title: 'Seventh',
    metric: '7d',
  },
  {
    title: 'Eighth',
    metric: '8d',
    disabled: true,
  },
  {
    title: 'Ninth',
    metric: '9d',
  },
  {
    title: 'Tenth',
    metric: '10d',
  },
];
