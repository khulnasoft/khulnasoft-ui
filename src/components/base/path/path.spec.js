import { nextTick } from 'vue';
import { shallowMount } from '@vue/test-utils';
import { mockPathItems } from './data';
import GlPath from './path.vue';

const PATH_ACTIVE_ITEM_CLASS = 'gl-path-active-item';
const BACKGROUND_COLOR_DEFAULT = 'rgba(0,0,0,0)';
const BACKGROUND_COLOR_LIGHT_GRAY = '#f0f0f0';

describe('Path', () => {
  let wrapper;

  const createComponent = (props = {}, options = {}) => {
    return shallowMount(GlPath, {
      propsData: {
        items: mockPathItems,
        ...props,
      },
      ...options,
    });
  };

  const pathNav = () => {
    return wrapper.find('[data-testid="gl-path-nav"]').element;
  };

  const listItems = () => {
    return wrapper.vm.$refs.pathListItems;
  };

  const pathItemAt = (index) => {
    return listItems()[index].children[0];
  };

  const pathItemTextAt = (index) => {
    return pathItemAt(index).textContent;
  };

  const clickItemAt = (index) => {
    pathItemAt(index).click();
  };

  beforeEach(() => {
    wrapper = createComponent();
  });

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('background color selection', () => {
    describe('with no background color specified', () => {
      it('displays the default background color', () => {
        expect(pathNav().style.getPropertyValue('--path-bg-color')).toBe(BACKGROUND_COLOR_DEFAULT);
      });
    });

    describe('with a background color specified', () => {
      beforeEach(() => {
        wrapper = createComponent({ backgroundColor: BACKGROUND_COLOR_LIGHT_GRAY });
      });

      it('sets the correct background color', () => {
        expect(pathNav().style.getPropertyValue('--path-bg-color')).toBe(
          BACKGROUND_COLOR_LIGHT_GRAY
        );
      });
    });
  });

  describe('renders the list of items', () => {
    it('renders the correct number of items', () => {
      expect(listItems().length).toBe(mockPathItems.length);
    });

    it('renders the items in the correct order', () => {
      expect(pathItemTextAt(0)).toContain(mockPathItems[0].title);
      expect(pathItemTextAt(4)).toContain(mockPathItems[4].title);
      expect(pathItemTextAt(9)).toContain(mockPathItems[9].title);
    });

    describe('with metrics', () => {
      beforeEach(() => {
        const data = mockPathItems;
        data[0].metric = '12d';

        wrapper = createComponent({ items: data });
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders the inline metric', () => {
        expect(pathItemTextAt(0)).toContain(mockPathItems[0].title);
        expect(pathItemTextAt(0)).toContain(mockPathItems[0].metric);
      });
    });

    describe('with icons', () => {
      const iconName = 'home';

      beforeEach(() => {
        const data = mockPathItems;
        data[0].icon = iconName;

        wrapper = createComponent({ items: data });
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders the inline icon', () => {
        const icon = wrapper.find('[data-testid="gl-path-item-icon"]');

        expect(icon.exists()).toBe(true);
        expect(icon.props('name')).toBe(iconName);
      });
    });
  });

  describe('renders the correct selected item', () => {
    describe('with no selected item passed in', () => {
      it('selects the first item', () => {
        expect(pathItemAt(0).classList).toContain(PATH_ACTIVE_ITEM_CLASS);
      });

      it('updates the selected item when props change', async () => {
        const items = JSON.parse(JSON.stringify(mockPathItems));
        items[3].selected = true;

        wrapper.setProps({ items });
        await nextTick();

        expect(pathItemAt(0).classList).not.toContain(PATH_ACTIVE_ITEM_CLASS);
        expect(pathItemAt(3).classList).toContain(PATH_ACTIVE_ITEM_CLASS);
      });
    });

    describe('with a specifically selected item passed in', () => {
      beforeEach(() => {
        const data = mockPathItems;
        data[3].selected = true;

        wrapper = createComponent({ items: data });
      });

      it('selects the correct item', () => {
        expect(pathItemAt(3).classList).toContain(PATH_ACTIVE_ITEM_CLASS);
      });
    });

    describe('with multiple selected items passed in', () => {
      beforeEach(() => {
        const data = mockPathItems;
        data[3].selected = true;
        data[5].selected = true;

        wrapper = createComponent({ items: data });
      });

      it('selects the first selected option', () => {
        expect(pathItemAt(3).classList).toContain(PATH_ACTIVE_ITEM_CLASS);
        expect(pathItemAt(5).classList).not.toContain(PATH_ACTIVE_ITEM_CLASS);
      });
    });
  });

  describe('event handling', () => {
    describe('when an item is clicked', () => {
      it('emits the selected event with the correct data', () => {
        clickItemAt(1);
        clickItemAt(4);
        clickItemAt(6);

        expect(wrapper.emitted('selected')).toEqual([
          [mockPathItems[1]],
          [mockPathItems[4]],
          [mockPathItems[6]],
        ]);
      });
    });

    describe('when a disabled item is clicked', () => {
      it('does not emit the selected event', () => {
        clickItemAt(7);

        expect(wrapper.emitted('selected')).toBeUndefined();
      });
    });
  });

  describe('slots', () => {
    beforeEach(() => {
      wrapper = createComponent(null, {
        scopedSlots: {
          default: `
            <div
              :data-pathid="props.pathId"
              data-testid="path-item-slot-content">
              {{ props.pathItem.title }}
            </div>
          `,
        },
      });
    });

    it('contains all elements passed into the default slot', () => {
      mockPathItems.forEach((item, index) => {
        const pathItem = wrapper.findAll('[data-testid="path-item-slot-content"]').at(index);

        expect(pathItem.text()).toBe(item.title);
        expect(pathItem.attributes('data-pathid')).toContain('path-');
      });
    });
  });
});
