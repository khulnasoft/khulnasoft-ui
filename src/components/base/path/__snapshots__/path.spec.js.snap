// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: rgba(0,0,0,0);"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-left"
        size="32"
        variant="current"
      />
    </button>
  </span>
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-0"
    >
      <button
        class="gl-path-button gl-path-active-item"
      >
        First
        <span
          class="gl-pl-2 gl-font-normal"
        >
          1d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-1"
    >
      <button
        class="gl-path-button"
      >
        Second
        <span
          class="gl-pl-2 gl-font-normal"
        >
          2d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-2"
    >
      <button
        class="gl-path-button"
      >
        Third
        <span
          class="gl-pl-2 gl-font-normal"
        >
          3d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-3"
    >
      <button
        class="gl-path-button"
      >
        Fourth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          4d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-4"
    >
      <button
        class="gl-path-button"
      >
        Fifth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          5d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-5"
    >
      <button
        class="gl-path-button"
      >
        Sixth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          6d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-6"
    >
      <button
        class="gl-path-button"
      >
        Seventh
        <span
          class="gl-pl-2 gl-font-normal"
        >
          7d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-7"
    >
      <button
        category="tertiary"
        class="gl-path-button"
        disabled=""
      >
        Eighth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          8d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-8"
    >
      <button
        class="gl-path-button"
      >
        Ninth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          9d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-1-item-9"
    >
      <button
        class="gl-path-button"
      >
        Tenth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          10d
        </span>
      </button>
    </li>
  </ul>
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-right"
        size="32"
        variant="current"
      />
    </button>
  </span>
</div>
`;

exports[`Path renders the list of items with icons matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: rgba(0,0,0,0);"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-left"
        size="32"
        variant="current"
      />
    </button>
  </span>
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-0"
    >
      <button
        class="gl-path-button gl-path-active-item"
      >
        <gl-icon-stub
          class="gl-mr-2"
          data-testid="gl-path-item-icon"
          name="home"
          size="16"
          variant="current"
        />
        First
        <span
          class="gl-pl-2 gl-font-normal"
        >
          12d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-1"
    >
      <button
        class="gl-path-button"
      >
        Second
        <span
          class="gl-pl-2 gl-font-normal"
        >
          2d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-2"
    >
      <button
        class="gl-path-button"
      >
        Third
        <span
          class="gl-pl-2 gl-font-normal"
        >
          3d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-3"
    >
      <button
        class="gl-path-button"
      >
        Fourth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          4d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-4"
    >
      <button
        class="gl-path-button"
      >
        Fifth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          5d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-5"
    >
      <button
        class="gl-path-button"
      >
        Sixth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          6d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-6"
    >
      <button
        class="gl-path-button"
      >
        Seventh
        <span
          class="gl-pl-2 gl-font-normal"
        >
          7d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-7"
    >
      <button
        category="tertiary"
        class="gl-path-button"
        disabled=""
      >
        Eighth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          8d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-8"
    >
      <button
        class="gl-path-button"
      >
        Ninth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          9d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-12-item-9"
    >
      <button
        class="gl-path-button"
      >
        Tenth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          10d
        </span>
      </button>
    </li>
  </ul>
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-right"
        size="32"
        variant="current"
      />
    </button>
  </span>
</div>
`;

exports[`Path renders the list of items with metrics matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: rgba(0,0,0,0);"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-left"
        size="32"
        variant="current"
      />
    </button>
  </span>
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-0"
    >
      <button
        class="gl-path-button gl-path-active-item"
      >
        First
        <span
          class="gl-pl-2 gl-font-normal"
        >
          12d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-1"
    >
      <button
        class="gl-path-button"
      >
        Second
        <span
          class="gl-pl-2 gl-font-normal"
        >
          2d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-2"
    >
      <button
        class="gl-path-button"
      >
        Third
        <span
          class="gl-pl-2 gl-font-normal"
        >
          3d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-3"
    >
      <button
        class="gl-path-button"
      >
        Fourth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          4d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-4"
    >
      <button
        class="gl-path-button"
      >
        Fifth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          5d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-5"
    >
      <button
        class="gl-path-button"
      >
        Sixth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          6d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-6"
    >
      <button
        class="gl-path-button"
      >
        Seventh
        <span
          class="gl-pl-2 gl-font-normal"
        >
          7d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-7"
    >
      <button
        category="tertiary"
        class="gl-path-button"
        disabled=""
      >
        Eighth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          8d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-8"
    >
      <button
        class="gl-path-button"
      >
        Ninth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          9d
        </span>
      </button>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-8-item-9"
    >
      <button
        class="gl-path-button"
      >
        Tenth
        <span
          class="gl-pl-2 gl-font-normal"
        >
          10d
        </span>
      </button>
    </li>
  </ul>
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <gl-icon-stub
        name="chevron-right"
        size="32"
        variant="current"
      />
    </button>
  </span>
</div>
`;
